### PR TITLE
add `handle_relations_with_same_arguments` parameter to `RETextClassificationWithIndicesTaskModule`

### PR DESCRIPTION
**Pull request by RainbowRivey**
If there are multiple relations with same pair of arguments, `handle_relations_with_same_arguments` defines if we remove both of them (`keep_none`) or keep the first one (`keep_first`). Full duplicates (if relation label is also the same) are not affected by this, one relation will be kept and a warning shown.

Note that if `collect_statistics=True`, final statistics do not include such "full duplicates" either as 'available' nor as 'skipped'. Also, **warnings about elements collected in the statistics**, e.g. skipped relations with same arguments, **will not be shown if `collect_statistics=True`.**
 
**This PR changes the default behavior to `keep_none`, that's why is labeled as `breaking` (previously it was implicitly `keep_first`).** Doing so prevents the model from learning conflicting relations and gaining biases for arbitrary 'first' relation occurred. 

TODO:
- [x] Test new behaviour on a downstream task (e.g. runs with 3 seeds at drugprot)

Note: `keep_none` reduces train instances count of drugprot by 199 (1,16%) from 17058 to 16859.
metric | baseline |  with keep_none 
-- | -- | -- 
quality (micro f1) | 0.7737 (±0.0042) |  0.7653 (±0.0022) 
Inference time | 13.5 (±0.3) | 11.2 (±0.5)

### CODE REVIEW

File: src/pie_modules/taskmodules/re_text_classification_with_indices.py [hunk 1]
```diff
@@ -337,6 +337,7 @@ def __init__(
         log_first_n_examples: int = 0,
         add_argument_indices_to_input: bool = False,
         add_global_attention_mask_to_input: bool = False,
+        relations_with_same_arguments: Optional[str] = "keep_none",
```
Comment (ArneBinder) [inline src/pie_modules/taskmodules/re_text_classification_with_indices.py:340]: rename to `handle_relations_with_same_arguments` (I know, a bit long, but much better to understand). Also, why `Optional`?

File: tests/taskmodules/test_re_text_classification_with_indices.py [hunk 2]
```diff
@@ -1042,10 +1042,12 @@ def test_collate_with_add_argument_indices(batch_with_argument_indices):
     torch.testing.assert_close(inputs["pooler_end_indices"], torch.tensor([[6, 11], [9, 14]]))
 
 
-def test_encode_input_multiple_relations_for_same_arguments(caplog):
+@pytest.mark.parametrize("relations_with_same_arguments", ["keep_first", "keep_none"])
+def test_encode_input_multiple_relations_for_same_arguments(caplog, relations_with_same_arguments):
```
Comment (ArneBinder) [inline tests/taskmodules/test_re_text_classification_with_indices.py:1046]: add one more relation `BinaryRelation(head=entities[0], tail=entities[1], label="per:founder")` to the document so that we can test the case where a relation is literally doubled (Github does not let me annotate the actual line below)

File: tests/taskmodules/test_re_text_classification_with_indices.py [hunk 3]
```diff
@@ -1042,10 +1042,12 @@ def test_collate_with_add_argument_indices(batch_with_argument_indices):
...
     taskmodule = RETextClassificationWithIndicesTaskModule(
         relation_annotation="relations",
         tokenizer_name_or_path="bert-base-cased",
+        relations_with_same_arguments=relations_with_same_arguments,
     )
```
Comment (ArneBinder) [inline tests/taskmodules/test_re_text_classification_with_indices.py:1051]: add `collect_statistics=True` and check the statistics

**Review by ArneBinder — CHANGES_REQUESTED** [review]
thanks, some comments below

EDIT: please also add some documentation about the new parameter to the docstring

File: tests/taskmodules/test_re_text_classification_with_indices.py (hunk 2 above)
Comment (RainbowRivey) [reply tests/taskmodules/test_re_text_classification_with_indices.py]: Added. There will be a warning, but no extra relation in statistics (statistics are somehow de-duplicated)

File: src/pie_modules/taskmodules/re_text_classification_with_indices.py (hunk 1 above)
Comment (RainbowRivey) [reply src/pie_modules/taskmodules/re_text_classification_with_indices.py]: No reason, fixed

File: tests/taskmodules/test_re_text_classification_with_indices.py (hunk 2 above)
Comment (ArneBinder) [reply tests/taskmodules/test_re_text_classification_with_indices.py]: ok. can you add a test case for that? i.e. that one of the duplicated relations is kept (independent of value of `handle_relations_with_same_arguments`)

Comment (ArneBinder) [reply tests/taskmodules/test_re_text_classification_with_indices.py]: regarding de-duplication in statistics: yeah, the concept of duplicated annotations does not really exist in pie. Maybe we can hack this by collecting the relation with a modified label? e.g.
```
self.collect_relation("skipped_same_arguments", rel.copy(label=f"{rel.label}_duplicate"))
```

**Comment by ArneBinder** [comment]
can you add a simple test for the not test-covered line? (from codecov report)

File: src/pie_modules/taskmodules/re_text_classification_with_indices.py [hunk 4]
```diff
@@ -693,11 +698,33 @@ def encode_input(
...
-                            f"doc.id={document.id}: there are multiple relations with the same arguments {arguments}: "
-                            f"previous label='{prev_label}' and current label='{rel.label}'. We only keep the first "
-                            f"occurring relation which has the label='{prev_label}'."
+                        arguments_resolved = tuple(
+                            map(lambda x: (x[0], x[1].resolve()), arguments)
                         )
+                        if prev_label == rel.label:
+                            logger.warning(
+                                f"doc.id={document.id}: Relation annotation `{rel.resolve()}` is duplicated. "
+                                f"We keep only one of them."
```
Comment (ArneBinder) [inline src/pie_modules/taskmodules/re_text_classification_with_indices.py:707]: mention that this does not reflect in the statistics (and how exactly)

File: src/pie_modules/taskmodules/re_text_classification_with_indices.py [hunk 5]
```diff
@@ -693,11 +698,33 @@ def encode_input(
...
+                        # if `keep_none`, first occurred relations are removed after _add_candidate_relations() call,
+                        # so that none of them are re-added as 'no-relation'
+                        elif self.handle_relations_with_same_arguments == "keep_none":
+                            logger.warning(
+                                f"doc.id={document.id}: there are multiple relations with the same arguments {arguments_resolved}: "
+                                f"previous label='{prev_label}' and current label='{rel.label}'. Both relations will be removed."
+                            )
+                        else:
+                            raise ValueError(
+                                f"'handle_relations_with_same_arguments' must be 'keep_first' or 'keep_none', but got `{self.handle_relations_with_same_arguments}`."
```
Comment (ArneBinder) [inline src/pie_modules/taskmodules/re_text_classification_with_indices.py:726]: shorten line

File: tests/taskmodules/test_re_text_classification_with_indices.py [hunk 6]
```diff
@@ -1058,26 +1065,74 @@ def test_encode_input_multiple_relations_for_same_arguments(caplog):
...
-        "and current label='per:founder'. We only keep the first occurring relation which has the "
-        "label='per:founded_by'."
-    )
-
-    assert len(encodings) == 1
-    relation = encodings[0].metadata["candidate_annotation"]
-    assert str(relation.head) == "A"
-    assert str(relation.tail) == "B"
-    assert relation.label == "per:founded_by"
+    if handle_relations_with_same_arguments not in ["keep_first", "keep_none"]:
```
Comment (ArneBinder) [inline tests/taskmodules/test_re_text_classification_with_indices.py:1073]:
```suggestion
    if handle_relations_with_same_arguments == "unknown_value":
```

File: tests/taskmodules/test_re_text_classification_with_indices.py [hunk 7]
```diff
@@ -1043,10 +1043,17 @@ def test_collate_with_add_argument_indices(batch_with_argument_indices):
     torch.testing.assert_close(inputs["pooler_end_indices"], torch.tensor([[6, 11], [9, 14]]))
 
 
-def test_encode_input_multiple_relations_for_same_arguments(caplog):
+@pytest.mark.parametrize(
+    "handle_relations_with_same_arguments", ["keep_first", "keep_none", "keep_second"]
+)
+def test_encode_input_multiple_relations_for_same_arguments(
```
Comment (ArneBinder) [inline tests/taskmodules/test_re_text_classification_with_indices.py:1049]: please add another test 
```
@pytest.mark.parametrize(
    "handle_relations_with_same_arguments", ["keep_first", "keep_none"]
)
def test_encode_input_duplicated_relations(caplog, handle_relations_with_same_arguments)
```
that tries a document with just one, but duplicated relation.

File: tests/taskmodules/test_re_text_classification_with_indices.py [hunk 8]
```diff
@@ -1058,26 +1065,74 @@ def test_encode_input_multiple_relations_for_same_arguments(caplog):
...
+            assert (
+                caplog.messages[2] == "statistics:\n"
+                "|                        |   per:founded_by |   per:founder |   all_relations |\n"
+                "|:-----------------------|-----------------:|--------------:|----------------:|\n"
+                "| available              |                1 |             1 |               2 |\n"
+                "| skipped_same_arguments |                0 |             1 |               1 |\n"
+                "| used                   |                1 |             0 |               1 |\n"
+                "| used %                 |              100 |             0 |              50 |"
+            )
+            assert len(encodings) == 1
```
Comment (ArneBinder) [inline tests/taskmodules/test_re_text_classification_with_indices.py:1109]: add a comment explaining this result (why one entry, given input doc with xxx?)

File: tests/taskmodules/test_re_text_classification_with_indices.py [hunk 9]
```diff
@@ -1058,26 +1065,74 @@ def test_encode_input_multiple_relations_for_same_arguments(caplog):
...
+                "We keep only one of them."
+            )
+            assert (
+                caplog.messages[2] == "statistics:\n"
+                "|                        |   per:founded_by |   per:founder |   all_relations |\n"
+                "|:-----------------------|-----------------:|--------------:|----------------:|\n"
+                "| available              |                1 |             1 |               2 |\n"
+                "| skipped_same_arguments |                1 |             1 |               2 |"
+            )
+            assert len(encodings) == 0
```
Comment (ArneBinder) [inline tests/taskmodules/test_re_text_classification_with_indices.py:1135]: add a comment explaining this result (why zero entries, given input doc with xxx?)

**Review by ArneBinder — CHANGES_REQUESTED** [review]
feedback below

File: src/pie_modules/taskmodules/re_text_classification_with_indices.py (hunk 4 above)
Comment (RainbowRivey) [reply src/pie_modules/taskmodules/re_text_classification_with_indices.py:707]: more detailed you mean?

Comment (ArneBinder) [reply src/pie_modules/taskmodules/re_text_classification_with_indices.py:707]: sorry, was not clear, I mean in this warning

**Review by ArneBinder — APPROVED** [review]
